src/services/refactors/convertReactPureComponent.ts

```diff
@@ -0,0 +1,294 @@
+     *        [ ] Provide an action to convert it to `React.SFC` or `React.PureComponent` (if it is `React.Component`)
+     * 4. [x] Resolve React by `getReferenceToReact()`
+     * 5. [ ] * Also support `propTypes`, `contextTypes`, `defaultProps`, `displayName` (Only Class -> Function)
+     * 6. [ ] * Also support convert of reference to `this.context` (in both ways)
+     */
+
+    registerRefactor(refactorName, {
+        getAvailableActions(context: RefactorContext): ApplicableRefactorInfo[] | undefined {
+            // if (!checkJSXFactoryIsReact(context.host)) return;
+            if (!getReferenceToReact(context.file)) return;
+            const node = getSelectedNode(context);
+            if (!node) { return; }
+
+            if (isConvertibleReactClassComponent(node, context.file)) {
+                // const description = Diagnostics.???
+                const description = actionNameComponentToSFC;
+                return [{
+                    name: refactorName,
+                    description,
+                    actions: [{ description, name: actionNameComponentToSFC }]
+                }];
+            }
+            else if (isReactSFCDeclaration(node, context.file)) {
+                // const description = Diagnostics.???
+                const description = actionNameSFCToPureComponent;
+                return [{
+                    name: refactorName,
+                    description,
+                    actions: [{ description, name: actionNameSFCToPureComponent }]
+                }];
+            }
+        },
+        getEditsForAction(context: RefactorContext, actionName: string): RefactorEditInfo | undefined {
+            Debug.assert(actionName === actionNameComponentToSFC || actionName === actionNameSFCToPureComponent);
+            const node = getSelectedNode(context) as Node;
+            if (actionName === actionNameSFCToPureComponent) {
+                return transformSFCtoComponent(isReactSFCDeclaration(node, context.file)!, context);
+            }
+            else if (actionName === actionNameComponentToSFC) {
+                return transformComponentToSFC(isConvertibleReactClassComponent(node, context.file)!, context);
+            }
+            return;
+        }
+    });
+
+    // function checkJSXFactoryIsReact(host: LanguageServiceHost) {
+    //     const config = host.getCompilationSettings();
+    //     if (config.jsxFactory === undefined) { if (config.jsx) return true; }
+    //     // In vue, people use jsxFactory: "h" insteadof React.createElement
+    //     else { if (config.jsxFactory.match(/React/)) return true; }
+    //     return false;
+    // }
+
+    function getSelectedNode(context: RefactorContext): Node | false {
+        const { file } = context;
+        const span = getRefactorContextSpan(context);
+        const token = getTokenAtPosition(file, span.start, /*includeJsDocComment*/ false);
+        const component = getParentNodeInSpan(token, file, span);

```

Comment:
this will enable the refactor in any location within the declaration. we want to limit it to spans that include the name of the declaration. so first get the token, walk up untill you find a class/class expression or a function/function expression/lambda, verify that the name of the class, or function keyword or class keyword or the name of the const/var on the left hand side is included in the span.

Reply:
see generateGetAccessorAndSetAccessor.ts:: getConvertibleFieldAtPosition for an example.